cou/steps/hypervisor.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2024 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Hypervisor planner."""
+import logging
+from dataclasses import dataclass
+
+from cou.apps.base import ApplicationUnit, OpenStackApplication
+from cou.steps import (
+    HypervisorUpgradePlan,
+    PostUpgradeStep,
+    PreUpgradeStep,
+    UpgradePlan,
+)
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Hypervisor:
+    """Hypervisor containing units for multiple applications."""
+
+    name: str
+    units: list[ApplicationUnit]
```

Comment:
Isn't this a machine in the end of the day? I'm not seeing much difference if the Machine class says which units are deployed on it. It seems that this Hypervisor is behaving like a "special" machine because it will have nova-compute on it.

Reply:
It's possible that we will use the machine class, I'm not sure until it will be used. This was done before Machine class exists and now, you are maybe right we do not need it anymore.

Reply:
I actually don't mind the separation between machine and hypervisor, but also have no objection on making them associated/inherited.